pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -214,40 +250,233 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+	for c.processNextClusterInfoImportItem() {
+	}
+}
+
+func (c *MCRouteController) processNextClusterInfoImportItem() bool {
+	obj, quit := c.ciImpQueue.Get()
 	if quit {
 		return false
 	}
-	defer c.queue.Done(obj)
+	defer c.ciImpQueue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
-		c.queue.Forget(obj)
+	ciImp, ok := obj.(string)
+	if !ok {
+		c.ciImpQueue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+	if c.wireGuardConfig != nil {
+		if err := c.updateWireGuardPeers(ciImp); err != nil {
+			c.ciImpQueue.AddRateLimited(ciImp)
+			klog.ErrorS(err, "Error updating WireGuard peers, requeuing", "key", ciImp)
+		}
+	}
+
+	if err := c.syncMCFlows(); err == nil {
+		c.ciImpQueue.Forget(obj)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.ciImpQueue.AddRateLimited(ciImp)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", ciImp)
+	}
+	return true
+}
+
+func (c *MCRouteController) processNextGatewayItem() bool {
+	obj, quit := c.gatewayQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gatewayQueue.Done(obj)
+
+	gateway, ok := obj.(string)
+	if !ok {
+		c.gatewayQueue.Forget(obj)
+		klog.InfoS("Expected string in work queue but got", "object", obj)
+		return true
+	}
+
+	if c.wireGuardConfig != nil {
+		if err := c.reconcileWireGuard(); err != nil {
+			klog.ErrorS(err, "Error when reconcile WireGuard, requeuing", "key", gateway)
+			c.gatewayQueue.AddRateLimited(gateway)
+		}
+	}
+	if err := c.syncMCFlows(); err == nil {
+		c.gatewayQueue.Forget(obj)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.gatewayQueue.AddRateLimited(gateway)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", gateway)
 	}
 	return true
 }
 
+func (c *MCRouteController) reconcileWireGuard() error {
+	gateways, err := c.mcClient.MulticlusterV1alpha1().Gateways(c.namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	if c.wireGuardInitialized {
+		// The Gateway is deleted. Remove WireGuard interface
+		if len(gateways.Items) == 0 || gateways.Items[0].Name != c.nodeConfig.Name {
+			klog.InfoS("Gateway is deleted, removing WireGuard interface and rule")
+			if err := c.deleteWireGuardInterfaceAndRule(); err != nil {
+				return err
+			}
+			c.wireGuardInitialized = false
+			return nil
+		}
+	}
+
+	// If the WireGuard is not initialized, and the Node is the Gateway Node.
+	// Controller will create corresponding interface and route on the host.
+	if !c.wireGuardInitialized && len(gateways.Items) != 0 && gateways.Items[0].Name == c.nodeConfig.Name {
+		klog.InfoS("Initializing WireGuard interface and rule")
+		if err := c.initializeWireGuard(); err != nil {
+			return err
+		}
+		if err := c.initWireGuardRoute(); err != nil {
```

Comment:
This one too if we move `updateWireGuardPeers()` into the if block.

Reply:
Moved